x/gov/keeper/proposal.go

```diff
@@ -55,15 +55,55 @@ func (keeper Keeper) GetProposal(ctx sdk.Context, proposalID uint64) (proposal t
 	if bz == nil {
 		return
 	}
-	keeper.cdc.MustUnmarshalBinaryLengthPrefixed(bz, &proposal)
+	if keeper.proposalCodecCtr != nil {
+		proposalI := keeper.proposalCodecCtr()
+		proto.Cdc.MustUnmarshalBinaryLengthPrefixed(bz, proposalI)
+		proposal = ProposalFromProposalI(proposalI)
+	} else {
+		keeper.cdc.MustUnmarshalBinaryLengthPrefixed(bz, proposal)
+	}
```

Comment:
@alexanderbez actually we could use this approach as an alternative.

Basically if there is a non-nil codec factory passed to the keeper use that and proposal will get marshaled to a protobuf `oneof`. If not, default to amino.

The advantage of this over https://github.com/cosmos/cosmos-sdk/pull/5533/files#diff-ecdb5afd213cbb6b18e4c9d1487c46ddR12-R20, is that the proposal codec is basically the proposal type generated by protobuf codegen. Just a thought